docs/source/design/cli.rst

```diff
@@ -0,0 +1,133 @@
+      directly in the command line. Its output is written to the ``stdout``,
+      ``stderr`` streams, and the exit code of the DataLad process matches
+      the exit code of the underlying command.
+
+   4. Keyboard interrupt (exit code 3)
+
+      The process was interrupted by the equivalent of a user hitting
+      ``Ctrl+C``.
+
+   5. Any other error/exception.
+
+
+Command parser construction by ``Interface`` inspection
+=======================================================
+
+The parser setup described here is implemented in ``datalad.cli.parser``.
+
+A dedicated sub-parser for any relevant DataLad command is constructed. For
+normal execution use cases, only a single subparser for the target command
+will be constructed for speed reasons. However, when the command line help
+system is requested (``--help``) subparsers for all commands (including
+extensions) are constructed. This can take a considerable amount of time
+that grows with the number of installed extensions.
+
+The information necessary to configure a subparser for a DataLad command is
+determine by inspecting the respective
+:class:`~datalad.interface.base.Interface` class for that command, and reusing
+individual components for the parser. This includes:
+
+- the class docstring
+
+- a ``_params_`` member with a dict of parameter definitions
+
+- a ``_examples_`` member, with a list of example definitions
+
+All docstrings used for the parser setup will be processed by applying a
+set of rules to make them more suitable for the command line environment.
+This includes the processing of ``CMD`` markup macros, and stripping their
+``PYTHON`` counter parts. Parameter constraint definition descriptions
+are also altered to exclude Python-specific idioms that have no relevance
+on the command line (e.g., the specification of ``None`` as a default).
+
+
+CLI-based execution of ``Interface`` command
+============================================
+
+The execution handler described here is implemented in ``datalad.cli.exec``.
+
+Once the main command line entry point determine that a command shall be
+executed, it triggers a handler function that was assigned and parameterized
+with the underlying command :class:`~datalad.interface.base.Interface` during
+parser construction. At the time of execution, this handler is given the result
+of ``argparsed``-based command line argument parsing (i.e., a ``Namespace``
+instance).
+
+From this parser result, the handler constructs positional and keyword
+arguments for the respective ``Interface.__call__()`` execution. It does
+not only process command-specific arguments, but also generic arguments,
+such as those for result filtering and rendering, which influence the central
+processing of result recorded yielded by a command.
+
+If an underlying command returns a Python generator it is unwound to trigger
+the respective underlying processing. The handler before no error handling.
+This is left to the main command line entry point.
```

Comment:
Some words appear to have been eaten (_The handler before no error handling._) but I can't figure out the desired statement.

Reply:
performs

Reply:
```suggestion
If an underlying command returns a Python generator it is unwound to trigger
the respective underlying processing. The handler performs no error handling.
This is left to the main command line entry point.
```